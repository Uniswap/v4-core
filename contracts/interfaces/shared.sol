// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.10;

/// @notice Represents a change in the pool's balance of token0 and token1.
struct BalanceDelta {
    int256 amount0;
    int256 amount1;
}
