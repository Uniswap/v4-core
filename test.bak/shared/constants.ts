export const MIN_TICK = -887272
export const MAX_TICK = 887272
export const MAX_TICK_SPACING = 32767
export const ADDRESS_ZERO = '0x0000000000000000000000000000000000000000'
