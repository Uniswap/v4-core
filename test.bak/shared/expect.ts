import { expect, use } from 'chai'
import { solidity } from 'ethereum-waffle'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'

use(solidity)
use(jestSnapshotPlugin())

export { expect }
