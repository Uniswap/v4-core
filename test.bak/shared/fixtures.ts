import { BigNumber } from 'ethers'
import { ethers } from 'hardhat'
import { TestERC20 } from '../../typechain/TestERC20'

interface TokensFixture {
  currency0: TestERC20
  currency1: TestERC20
  token2: TestERC20
}

export async function tokensFixture(): Promise<TokensFixture> {
  const tokenFactory = await ethers.getContractFactory('TestERC20')
  const tokenA = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20
  const tokenB = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20
  const tokenC = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20

  const [currency0, currency1, token2] = [tokenA, tokenB, tokenC].sort((tokenA, tokenB) =>
    tokenA.address.toLowerCase() < tokenB.address.toLowerCase() ? -1 : 1
  )

  return { currency0, currency1, token2 }
}
