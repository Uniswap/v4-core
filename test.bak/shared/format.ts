import { Decimal } from 'decimal.js'
import { BigNumberish } from 'ethers'

export function formatTokenAmount(num: BigNumberish): string {
  return new Decimal(num.toString()).dividedBy(new Decimal(10).pow(18)).toPrecision(5)
}

export function formatPrice(price: BigNumberish): string {
  return new Decimal(price.toString()).dividedBy(new Decimal(2).pow(96)).pow(2).toPrecision(5)
}
