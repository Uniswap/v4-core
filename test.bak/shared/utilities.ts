import bn from 'bignumber.js'
import { BigNumber, BigNumberish, Contract, ContractTransaction, utils, Wallet } from 'ethers'
import { ethers } from 'hardhat'

export const MaxUint128 = BigNumber.from(2).pow(128).sub(1)

export const getMinTick = (tickSpacing: number) => Math.ceil(-887272 / tickSpacing) * tickSpacing
export const getMaxTick = (tickSpacing: number) => Math.floor(887272 / tickSpacing) * tickSpacing

export const MIN_SQRT_RATIO = BigNumber.from('4295128739')
export const MAX_SQRT_RATIO = BigNumber.from('1461446703485210103287273052203988822378723970342')

export enum FeeAmount {
  LOW = 500,
  MEDIUM = 3000,
  HIGH = 10000,
}

export const TICK_SPACINGS: { [amount in FeeAmount]: number } = {
  [FeeAmount.LOW]: 10,
  [FeeAmount.MEDIUM]: 60,
  [FeeAmount.HIGH]: 200,
}

export function expandTo18Decimals(n: number): BigNumber {
  return BigNumber.from(n).mul(BigNumber.from(10).pow(18))
}

bn.config({ EXPONENTIAL_AT: 999999, DECIMAL_PLACES: 40 })

// returns the sqrt price as a 64x96
export function encodeSqrtPriceX96(reserve1: BigNumberish, reserve0: BigNumberish): BigNumber {
  return BigNumber.from(
    new bn(reserve1.toString())
      .div(reserve0.toString())
      .sqrt()
      .multipliedBy(new bn(2).pow(96))
      .integerValue(3)
      .toString()
  )
}

export function getPositionKey(address: string, lowerTick: number, upperTick: number): string {
  return utils.keccak256(utils.solidityPack(['address', 'int24', 'int24'], [address, lowerTick, upperTick]))
}

export function getPoolId({
  currency0,
  currency1,
  fee,
  tickSpacing,
  hooks,
}: {
  currency0: string | Contract
  currency1: string | Contract
  fee: number
  tickSpacing: number
  hooks: string | Contract
}): string {
  return utils.keccak256(
    utils.defaultAbiCoder.encode(
      ['address', 'address', 'uint24', 'int24', 'address'],
      [
        typeof currency0 === 'string' ? currency0 : currency0.address,
        typeof currency1 === 'string' ? currency1 : currency1.address,
        fee,
        tickSpacing,
        typeof hooks === 'string' ? hooks : hooks.address,
      ]
    )
  )
}

export type SwapFunction = (
  amount: BigNumberish,
  to: Wallet | string,
  sqrtPriceLimitX96?: BigNumberish
) => Promise<ContractTransaction>
export type SwapToPriceFunction = (sqrtPriceX96: BigNumberish, to: Wallet | string) => Promise<ContractTransaction>
export type ModifyPositionFunction = (
  tickLower: BigNumberish,
  tickUpper: BigNumberish,
  liquidityDelta: BigNumberish
) => Promise<ContractTransaction>
export type DonateFunction = (amount0: BigNumberish, amount1: BigNumberish) => Promise<ContractTransaction>

interface HookMask {
  beforeInitialize: boolean
  afterInitialize: boolean
  beforeModifyPosition: boolean
  afterModifyPosition: boolean
  beforeSwap: boolean
  afterSwap: boolean
  beforeDonate: boolean
  afterDonate: boolean
}

/**
 * Creates a 20 byte mask for the given hook configuration
 */
export function createHookMask({
  beforeInitialize,
  afterInitialize,
  beforeModifyPosition,
  afterModifyPosition,
  beforeSwap,
  afterSwap,
  beforeDonate,
  afterDonate,
}: HookMask): string {
  let result: BigNumber = BigNumber.from(0)
  if (beforeInitialize) result = result.add(BigNumber.from(1).shl(159))
  if (afterInitialize) result = result.add(BigNumber.from(1).shl(158))
  if (beforeModifyPosition) result = result.add(BigNumber.from(1).shl(157))
  if (afterModifyPosition) result = result.add(BigNumber.from(1).shl(156))
  if (beforeSwap) result = result.add(BigNumber.from(1).shl(155))
  if (afterSwap) result = result.add(BigNumber.from(1).shl(154))
  if (beforeDonate) result = result.add(BigNumber.from(1).shl(153))
  if (afterDonate) result = result.add(BigNumber.from(1).shl(152))
  return utils.hexZeroPad(result.toHexString(), 20)
}

/**
 * Returns a wallet whose first transaction will create a contract satisfying the leading
 * bytes required by hookMask. If provided, the mnemonic argument short-circuits our search
 * to save time.
 */
export function getWalletForDeployingHookMask(hookMask: HookMask, mnemonic?: string): [Wallet, string] {
  const startingString = createHookMask(hookMask).slice(0, 4)
  let wallet: Wallet = mnemonic ? ethers.Wallet.fromMnemonic(mnemonic) : ethers.Wallet.createRandom()
  let contractAddress: string | undefined

  while (contractAddress === undefined) {
    const prospectiveContractAddress = utils.getContractAddress({ from: wallet.address, nonce: 0 })
    if (prospectiveContractAddress.slice(0, 4).toLowerCase() === startingString) {
      contractAddress = prospectiveContractAddress
    } else {
      // if, for whatever reason, we generate a bad address but a mnemonic was provided,
      // it's stale and we surface the error
      if (mnemonic) throw Error('Stale mnemonic')
      wallet = ethers.Wallet.createRandom()
    }
  }

  return [wallet, contractAddress]
}
