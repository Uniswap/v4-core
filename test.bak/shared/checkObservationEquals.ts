import { BigNumber, BigNumberish } from 'ethers'
import { expect } from './expect'

// helper function because we cannot do a simple deep equals with the
// observation result object returned from ethers because it extends array
export default function checkObservationEquals(
  {
    tickCumulative,
    blockTimestamp,
    initialized,
    secondsPerLiquidityCumulativeX128,
  }: {
    tickCumulative: BigNumber
    secondsPerLiquidityCumulativeX128: BigNumber
    initialized: boolean
    blockTimestamp: number
  },
  expected: {
    tickCumulative: BigNumberish
    secondsPerLiquidityCumulativeX128: BigNumberish
    initialized: boolean
    blockTimestamp: number
  }
) {
  expect(
    {
      initialized,
      blockTimestamp,
      tickCumulative: tickCumulative.toString(),
      secondsPerLiquidityCumulativeX128: secondsPerLiquidityCumulativeX128.toString(),
    },
    `observation is equivalent`
  ).to.deep.eq({
    ...expected,
    tickCumulative: expected.tickCumulative.toString(),
    secondsPerLiquidityCumulativeX128: expected.secondsPerLiquidityCumulativeX128.toString(),
  })
}
