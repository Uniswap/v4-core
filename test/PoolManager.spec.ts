import snapshotGasCost from '@uniswap/snapshot-gas-cost'
import { BigNumber, Wallet } from 'ethers'
import hre, { ethers, waffle } from 'hardhat'
import {
  EmptyTestHooks,
  PoolDonateTest,
  PoolManager,
  PoolManagerReentrancyTest,
  PoolModifyPositionTest,
  PoolSwapTest,
  PoolTakeTest,
  ProtocolFeeControllerTest,
  TestERC20,
} from '../typechain'
import { ADDRESS_ZERO, MAX_TICK_SPACING } from './shared/constants'
import { expect } from './shared/expect'
import { tokensFixture } from './shared/fixtures'
import { MockedContract, deployMockContract, setCode } from './shared/mockContract'
import { FeeAmount, MaxUint128, encodeSqrtPriceX96, expandTo18Decimals, getPoolId } from './shared/utilities'

const createFixtureLoader = waffle.createFixtureLoader

const { constants } = ethers

describe('PoolManager', () => {
  let wallet: Wallet, other: Wallet

  let manager: PoolManager
  let swapTest: PoolSwapTest
  let feeControllerTest: ProtocolFeeControllerTest
  let modifyPositionTest: PoolModifyPositionTest
  let donateTest: PoolDonateTest
  let takeTest: PoolTakeTest
  let hooksMock: MockedContract
  let testHooksEmpty: EmptyTestHooks
  let tokens: { currency0: TestERC20; currency1: TestERC20; token2: TestERC20 }

  const fixture = async () => {
    const poolManagerFactory = await ethers.getContractFactory('PoolManager')
    const swapTestFactory = await ethers.getContractFactory('PoolSwapTest')
    const feeControllerTestFactory = await ethers.getContractFactory('ProtocolFeeControllerTest')
    const modifyPositionTestFactory = await ethers.getContractFactory('PoolModifyPositionTest')
    const donateTestFactory = await ethers.getContractFactory('PoolDonateTest')
    const takeTestFactory = await ethers.getContractFactory('PoolTakeTest')
    const hooksTestEmptyFactory = await ethers.getContractFactory('EmptyTestHooks')
    const tokens = await tokensFixture()
    const CONTROLLER_GAS_LIMIT = 50000
    const manager = (await poolManagerFactory.deploy(CONTROLLER_GAS_LIMIT)) as PoolManager

    // Deploy hooks to addresses with leading 1111 to enable all of them.
    const mockHooksAddress = '0xFF00000000000000000000000000000000000000'
    const testHooksEmptyAddress = '0xF000000000000000000000000000000000000000'

    const hookImplAddress = '0xFF00000000000000000000000000000000000001'
    setCode(hookImplAddress, 'EmptyTestHooks')
    hooksMock = await deployMockContract(hooksTestEmptyFactory.interface, mockHooksAddress, hookImplAddress)

    await hre.network.provider.send('hardhat_setCode', [
      testHooksEmptyAddress,
      (await hre.artifacts.readArtifact('EmptyTestHooks')).deployedBytecode,
    ])

    const testHooksEmpty: EmptyTestHooks = hooksTestEmptyFactory.attach(testHooksEmptyAddress) as EmptyTestHooks

    const result = {
      manager,
      swapTest: (await swapTestFactory.deploy(manager.address)) as PoolSwapTest,
      feeControllerTest: (await feeControllerTestFactory.deploy()) as ProtocolFeeControllerTest,
      modifyPositionTest: (await modifyPositionTestFactory.deploy(manager.address)) as PoolModifyPositionTest,
      donateTest: (await donateTestFactory.deploy(manager.address)) as PoolDonateTest,
      takeTest: (await takeTestFactory.deploy(manager.address)) as PoolTakeTest,
      tokens,
      hooksMock,
      testHooksEmpty,
    }

    for (const token of [tokens.currency0, tokens.currency1, tokens.token2]) {
      for (const spender of [result.swapTest, result.modifyPositionTest, result.donateTest, result.takeTest]) {
        await token.connect(wallet).approve(spender.address, constants.MaxUint256)
      }
    }

    return result
  }

  let loadFixture: ReturnType<typeof createFixtureLoader>
  before('create fixture loader', async () => {
    ;[wallet, other] = await (ethers as any).getSigners()

    loadFixture = createFixtureLoader([wallet, other])
  })

  beforeEach('deploy fixture', async () => {
    ;({
      manager,
      tokens,
      modifyPositionTest,
      swapTest,
      feeControllerTest,
      donateTest,
      takeTest,
      hooksMock,
      testHooksEmpty,
    } = await loadFixture(fixture))
  })

  it('bytecode size', async () => {
    expect(((await waffle.provider.getCode(manager.address)).length - 2) / 2).to.matchSnapshot()
  })

  describe('#lock', () => {
    it('can be reentered', async () => {
      const reenterTest = (await (
        await ethers.getContractFactory('PoolManagerReentrancyTest')
      ).deploy()) as PoolManagerReentrancyTest

      await manager.initialize(
        {
          currency0: tokens.currency0.address,
          currency1: tokens.currency1.address,
          fee: FeeAmount.MEDIUM,
          tickSpacing: 60,
          hooks: ADDRESS_ZERO,
        },
        encodeSqrtPriceX96(1, 1)
      )

      await modifyPositionTest.modifyPosition(
        {
          currency0: tokens.currency0.address,
          currency1: tokens.currency1.address,
          fee: FeeAmount.MEDIUM,
          tickSpacing: 60,
          hooks: ADDRESS_ZERO,
        },
        {
          tickLower: 0,
          tickUpper: 60,
          liquidityDelta: 100,
        }
      )

      await expect(reenterTest.reenter(manager.address, tokens.currency0.address, 3))
        .to.emit(reenterTest, 'LockAcquired')
        .withArgs(3)
        .to.emit(reenterTest, 'LockAcquired')
        .withArgs(2)
        .to.emit(reenterTest, 'LockAcquired')
        .withArgs(1)
    })
  })

  describe('#setProtocolFeeController', () => {
    it('allows the owner to set a fee controller', async () => {
      expect(await manager.protocolFeeController()).to.be.eq(ADDRESS_ZERO)
      await expect(manager.setProtocolFeeController(feeControllerTest.address)).to.emit(
        manager,
        'ProtocolFeeControllerUpdated'
      )
      expect(await manager.protocolFeeController()).to.be.eq(feeControllerTest.address)
    })
  })

  describe('#take', () => {
    it('fails if no liquidity', async () => {
      await tokens.currency0.connect(wallet).transfer(ADDRESS_ZERO, constants.MaxUint256.div(2))
      await expect(
        takeTest.connect(wallet).take(
          {
            currency0: tokens.currency0.address,
            currency1: tokens.currency1.address,
            fee: 100,
            hooks: ADDRESS_ZERO,
            tickSpacing: 10,
          },
          100,
          0
        )
      ).to.be.reverted
    })

    it('fails for invalid tokens that dont return true on transfer', async () => {
      const tokenFactory = await ethers.getContractFactory('TestInvalidERC20')
      const invalidToken = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20
      const currency0Invalid = invalidToken.address.toLowerCase() < tokens.currency0.address.toLowerCase()
      const key = {
        currency0: currency0Invalid ? invalidToken.address : tokens.currency0.address,
        currency1: currency0Invalid ? tokens.currency0.address : invalidToken.address,
        fee: 100,
        hooks: ADDRESS_ZERO,
        tickSpacing: 10,
      }
      await invalidToken.approve(modifyPositionTest.address, constants.MaxUint256)
      await manager.initialize(key, encodeSqrtPriceX96(1, 1))
      await modifyPositionTest.modifyPosition(key, {
        tickLower: -60,
        tickUpper: 60,
        liquidityDelta: 100,
      })

      await tokens.currency0.connect(wallet).approve(takeTest.address, MaxUint128)
      await invalidToken.connect(wallet).approve(takeTest.address, MaxUint128)

      await expect(takeTest.connect(wallet).take(key, currency0Invalid ? 1 : 0, currency0Invalid ? 0 : 1)).to.be
        .reverted
      await expect(takeTest.connect(wallet).take(key, currency0Invalid ? 0 : 1, currency0Invalid ? 1 : 0)).to.not.be
        .reverted
    })

    it('succeeds if has liquidity', async () => {
      const key = {
        currency0: tokens.currency0.address,
        currency1: tokens.currency1.address,
        fee: 100,
        hooks: ADDRESS_ZERO,
        tickSpacing: 10,
      }
      await manager.initialize(key, encodeSqrtPriceX96(1, 1))
      await modifyPositionTest.modifyPosition(key, {
        tickLower: -60,
        tickUpper: 60,
        liquidityDelta: 100,
      })

      await tokens.currency0.connect(wallet).approve(takeTest.address, MaxUint128)

      await expect(takeTest.connect(wallet).take(key, 1, 0)).to.be.not.be.reverted
      await expect(takeTest.connect(wallet).take(key, 0, 1)).to.be.not.be.reverted
    })

    it('succeeds with native tokens if has liquidity', async () => {
      const key = {
        currency0: ADDRESS_ZERO,
        currency1: tokens.currency1.address,
        fee: 100,
        hooks: ADDRESS_ZERO,
        tickSpacing: 10,
      }
      await manager.initialize(key, encodeSqrtPriceX96(1, 1))
      await modifyPositionTest.modifyPosition(
        key,
        {
          tickLower: -60,
          tickUpper: 60,
          liquidityDelta: 100,
        },
        { value: 100 }
      )

      await tokens.currency0.connect(wallet).approve(takeTest.address, MaxUint128)

      await expect(takeTest.connect(wallet).take(key, 1, 0, { value: 1 })).to.be.not.be.reverted
      await expect(takeTest.connect(wallet).take(key, 0, 1)).to.be.not.be.reverted
    })
  })

  describe('#setPoolProtocolFee', async () => {
    it('updates the protocol fee for an initialised pool', async () => {
      expect(await manager.protocolFeeController()).to.be.eq(ADDRESS_ZERO)

      const poolKey = {
        currency0: tokens.currency0.address,
        currency1: tokens.currency1.address,
        fee: FeeAmount.MEDIUM,
        tickSpacing: 60,
        hooks: ADDRESS_ZERO,
      }
      const poolID = getPoolId(poolKey)

      await manager.initialize(poolKey, encodeSqrtPriceX96(10, 1))

      var protocolSwapFee: number
      ;({
        slot0: { protocolSwapFee },
      } = await manager.pools(getPoolId(poolKey)))
      expect(protocolSwapFee).to.eq(0)

      await manager.setProtocolFeeController(feeControllerTest.address)
      expect(await manager.protocolFeeController()).to.be.eq(feeControllerTest.address)
      const poolProtocolFee = 4
      await feeControllerTest.setSwapFeeForPool(poolID, poolProtocolFee)

      await expect(manager.setProtocolFees(poolKey))
        .to.emit(manager, 'ProtocolFeeUpdated')
        .withArgs(poolID, poolProtocolFee, 0)
      ;({
        slot0: { protocolSwapFee },
      } = await manager.pools(poolID))
      expect(protocolSwapFee).to.eq(poolProtocolFee)
    })
  })

  describe('#collectProtocolFees', async () => {
    describe('ERC20 tokens', async () => {
      beforeEach('set the fee controller, initialize a pool with protocol fee', async () => {
        const poolKey = {
          currency0: tokens.currency0.address,
          currency1: tokens.currency1.address,
          fee: FeeAmount.MEDIUM,
          tickSpacing: 60,
          hooks: ADDRESS_ZERO,
        }
        // set the controller, and set the pool's protocol fee
        await manager.setProtocolFeeController(feeControllerTest.address)
        expect(await manager.protocolFeeController()).to.be.eq(feeControllerTest.address)
        const poolProtocolFee = 68 // 0x 0100 0100
        const poolID = getPoolId(poolKey)
        await feeControllerTest.setSwapFeeForPool(poolID, poolProtocolFee)

        // initialize the pool with the fee
        await manager.initialize(poolKey, encodeSqrtPriceX96(1, 1))
        const {
          slot0: { protocolSwapFee },
        } = await manager.pools(getPoolId(poolKey))
        expect(protocolSwapFee).to.eq(poolProtocolFee)

        // add liquidity around the initial price
        await modifyPositionTest.modifyPosition(poolKey, {
          tickLower: -120,
          tickUpper: 120,
          liquidityDelta: expandTo18Decimals(10),
        })
      })

      it('allows the owner to collect accumulated fees', async () => {
        await swapTest.swap(
          {
            currency0: tokens.currency0.address,
            currency1: tokens.currency1.address,
            fee: FeeAmount.MEDIUM,
            tickSpacing: 60,
            hooks: ADDRESS_ZERO,
          },
          {
            amountSpecified: 10000,
            sqrtPriceLimitX96: encodeSqrtPriceX96(1, 2),
            zeroForOne: true,
          },
          {
            withdrawTokens: true,
            settleUsingTransfer: true,
          }
        )

        const expectedFees = 7
        expect(await manager.protocolFeesAccrued(tokens.currency0.address)).to.be.eq(BigNumber.from(expectedFees))
        expect(await manager.protocolFeesAccrued(tokens.currency1.address)).to.be.eq(BigNumber.from(0))

        // allows the owner to collect the fees
        const recipientBalanceBefore = await tokens.currency0.balanceOf(other.address)
        const managerBalanceBefore = await tokens.currency0.balanceOf(manager.address)

        // get the returned value, then actually execute
        const amount = await manager.callStatic.collectProtocolFees(other.address, tokens.currency0.address, 7)
        await manager.collectProtocolFees(other.address, tokens.currency0.address, expectedFees)

        const recipientBalanceAfter = await tokens.currency0.balanceOf(other.address)
        const managerBalanceAfter = await tokens.currency0.balanceOf(manager.address)

        expect(amount).to.be.eq(expectedFees)
        expect(recipientBalanceAfter).to.be.eq(recipientBalanceBefore.add(expectedFees))
        expect(managerBalanceAfter).to.be.eq(managerBalanceBefore.sub(expectedFees))

        expect(await manager.protocolFeesAccrued(tokens.currency0.address)).to.be.eq(BigNumber.from(0))
      })

      it('returns all fees if 0 is provided', async () => {
        await swapTest.swap(
          {
            currency0: tokens.currency0.address,
            currency1: tokens.currency1.address,
            fee: FeeAmount.MEDIUM,
            tickSpacing: 60,
            hooks: ADDRESS_ZERO,
          },
          {
            amountSpecified: 10000,
            sqrtPriceLimitX96: encodeSqrtPriceX96(1, 2),
            zeroForOne: true,
          },
          {
            withdrawTokens: true,
            settleUsingTransfer: true,
          }
        )

        const expectedFees = 7
        expect(await manager.protocolFeesAccrued(tokens.currency0.address)).to.be.eq(BigNumber.from(expectedFees))
        expect(await manager.protocolFeesAccrued(tokens.currency1.address)).to.be.eq(BigNumber.from(0))

        // allows the owner to collect the fees
        const recipientBalanceBefore = await tokens.currency0.balanceOf(other.address)
        const managerBalanceBefore = await tokens.currency0.balanceOf(manager.address)

        // get the returned value, then actually execute
        const amount = await manager.callStatic.collectProtocolFees(other.address, tokens.currency0.address, 0)
        await manager.collectProtocolFees(other.address, tokens.currency0.address, 0)

        const recipientBalanceAfter = await tokens.currency0.balanceOf(other.address)
        const managerBalanceAfter = await tokens.currency0.balanceOf(manager.address)

        expect(amount).to.be.eq(expectedFees)
        expect(recipientBalanceAfter).to.be.eq(recipientBalanceBefore.add(expectedFees))
        expect(managerBalanceAfter).to.be.eq(managerBalanceBefore.sub(expectedFees))

        expect(await manager.protocolFeesAccrued(tokens.currency0.address)).to.be.eq(BigNumber.from(0))
      })
    })

    describe('native tokens', async () => {
      beforeEach('set the fee controller, initialize a pool with protocol fee', async () => {
        const poolKey = {
          currency0: ADDRESS_ZERO,
          currency1: tokens.currency1.address,
          fee: FeeAmount.MEDIUM,
          tickSpacing: 60,
          hooks: ADDRESS_ZERO,
        }
        // set the controller, and set the pool's protocol fee
        await manager.setProtocolFeeController(feeControllerTest.address)
        expect(await manager.protocolFeeController()).to.be.eq(feeControllerTest.address)
        const poolProtocolFee = 68 // 0x 0100 0100
        const poolID = getPoolId(poolKey)
        await feeControllerTest.setSwapFeeForPool(poolID, poolProtocolFee)

        // initialize the pool with the fee
        await manager.initialize(poolKey, encodeSqrtPriceX96(1, 1))
        const {
          slot0: { protocolSwapFee },
        } = await manager.pools(getPoolId(poolKey))
        expect(protocolSwapFee).to.eq(poolProtocolFee)

        // add liquidity around the initial price
        await modifyPositionTest.modifyPosition(
          poolKey,
          {
            tickLower: -120,
            tickUpper: 120,
            liquidityDelta: expandTo18Decimals(10),
          },
          {
            value: expandTo18Decimals(10),
          }
        )
      })

      it('allows the owner to collect accumulated fees', async () => {
        await swapTest.swap(
          {
            currency0: ADDRESS_ZERO,
            currency1: tokens.currency1.address,
            fee: FeeAmount.MEDIUM,
            tickSpacing: 60,
            hooks: ADDRESS_ZERO,
          },
          {
            amountSpecified: 10000,
            sqrtPriceLimitX96: encodeSqrtPriceX96(1, 2),
            zeroForOne: true,
          },
          {
            withdrawTokens: true,
            settleUsingTransfer: true,
          },
          {
            value: 10000,
          }
        )

        const expectedFees = 7
        expect(await manager.protocolFeesAccrued(ADDRESS_ZERO)).to.be.eq(BigNumber.from(expectedFees))
        expect(await manager.protocolFeesAccrued(tokens.currency1.address)).to.be.eq(BigNumber.from(0))

        // allows the owner to collect the fees
        const recipientBalanceBefore = await other.getBalance()
        const managerBalanceBefore = await waffle.provider.getBalance(manager.address)

        // get the returned value, then actually execute
        const amount = await manager.callStatic.collectProtocolFees(other.address, ADDRESS_ZERO, 7)
        await manager.collectProtocolFees(other.address, ADDRESS_ZERO, expectedFees)

        const recipientBalanceAfter = await other.getBalance()
        const managerBalanceAfter = await waffle.provider.getBalance(manager.address)

        expect(amount).to.be.eq(expectedFees)
        expect(recipientBalanceAfter).to.be.eq(recipientBalanceBefore.add(expectedFees))
        expect(managerBalanceAfter).to.be.eq(managerBalanceBefore.sub(expectedFees))

        expect(await manager.protocolFeesAccrued(ADDRESS_ZERO)).to.be.eq(BigNumber.from(0))
      })

      it('returns all fees if 0 is provided', async () => {
        await swapTest.swap(
          {
            currency0: ADDRESS_ZERO,
            currency1: tokens.currency1.address,
            fee: FeeAmount.MEDIUM,
            tickSpacing: 60,
            hooks: ADDRESS_ZERO,
          },
          {
            amountSpecified: 10000,
            sqrtPriceLimitX96: encodeSqrtPriceX96(1, 2),
            zeroForOne: true,
          },
          {
            withdrawTokens: true,
            settleUsingTransfer: true,
          },
          {
            value: 10000,
          }
        )

        const expectedFees = 7
        expect(await manager.protocolFeesAccrued(ADDRESS_ZERO)).to.be.eq(BigNumber.from(expectedFees))
        expect(await manager.protocolFeesAccrued(tokens.currency1.address)).to.be.eq(BigNumber.from(0))

        // allows the owner to collect the fees
        const recipientBalanceBefore = await other.getBalance()
        const managerBalanceBefore = await waffle.provider.getBalance(manager.address)

        // get the returned value, then actually execute
        const amount = await manager.callStatic.collectProtocolFees(other.address, ADDRESS_ZERO, 0)
        await manager.collectProtocolFees(other.address, ADDRESS_ZERO, 0)

        const recipientBalanceAfter = await other.getBalance()
        const managerBalanceAfter = await waffle.provider.getBalance(manager.address)

        expect(amount).to.be.eq(expectedFees)
        expect(recipientBalanceAfter).to.be.eq(recipientBalanceBefore.add(expectedFees))
        expect(managerBalanceAfter).to.be.eq(managerBalanceBefore.sub(expectedFees))

        expect(await manager.protocolFeesAccrued(ADDRESS_ZERO)).to.be.eq(BigNumber.from(0))
      })
    })
  })
})
