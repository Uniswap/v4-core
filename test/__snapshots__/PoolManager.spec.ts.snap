// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PoolManager #initialize gas cost 1`] = `
Object {
  "calldataByteLength": 196,
  "gasUsed": 49105,
}
`;

exports[`PoolManager #lock gas overhead of no-op lock 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 59780,
}
`;

exports[`PoolManager #mint gas cost 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 287310,
}
`;

exports[`PoolManager #mint gas cost with hooks 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 291426,
}
`;

exports[`PoolManager #swap gas cost 1`] = `
Object {
  "calldataByteLength": 324,
  "gasUsed": 84550,
}
`;

exports[`PoolManager #swap gas cost for swap against liquidity 1`] = `
Object {
  "calldataByteLength": 324,
  "gasUsed": 224036,
}
`;

exports[`PoolManager #swap gas cost with hooks 1`] = `
Object {
  "calldataByteLength": 324,
  "gasUsed": 84579,
}
`;

exports[`PoolManager bytecode size 1`] = `23362`;
